File: src/main/java/io/github/colintimbarndt/chat_emotes/command/ChatEmotesCommand.java
package io.github.colintimbarndt.chat_emotes.command;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.context.CommandContext;
import io.github.colintimbarndt.chat_emotes.ChatEmotesMod;
import static net.minecraft.server.command.CommandManager.literal;

import static io.github.colintimbarndt.chat_emotes.TranslationHelper.translatable;
import net.minecraft.server.command.ServerCommandSource;

public class ChatEmotesCommand {
    private static int tryReloadConfig(CommandContext<ServerCommandSource> context) {
        final var source = context.getSource();
        if (ChatEmotesMod.reloadConfig()) {
            source.sendFeedback(
                    translatable("chat_emotes.commands.reload.success"),
                    true
            );
            return 1;
        } else {
            source.sendFeedback(
                    translatable("chat_emotes.commands.reload.failure"),
                    true
            );
            return 0;
        }
    }

    public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
        dispatcher.register(
                literal("chat-emotes")
                        .then(
                                literal("reload")
                                        .requires(source -> source.hasPermissionLevel(2))
                                        .executes(ChatEmotesCommand::tryReloadConfig)
                        )
        );
    }
}

File: mod/common_mod_server/src/main/kotlin/io/github/colintimbarndt/chat_emotes/command/ChatEmotesCommand.kt
package io.github.colintimbarndt.chat_emotes.command

import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.context.CommandContext
import io.github.colintimbarndt.chat_emotes.ComponentFactory
import io.github.colintimbarndt.chat_emotes.common.ChatEmotesServerModBase
import io.github.colintimbarndt.chat_emotes.common.NAMESPACE
import io.github.colintimbarndt.chat_emotes.common.commands.ChatEmotesCommandBase
import io.github.colintimbarndt.chat_emotes.common.permissions.COMMAND_PERMISSION
import io.github.colintimbarndt.chat_emotes.common.permissions.RELOAD_COMMAND_PERMISSION
import io.github.colintimbarndt.chat_emotes.common.permissions.permissionPredicate
import net.minecraft.commands.CommandSourceStack
import net.minecraft.commands.Commands
import net.minecraft.network.chat.Component

private typealias Ctx = CommandContext<CommandSourceStack>

class ChatEmotesCommand(
    private val mod: ChatEmotesServerModBase
) : ChatEmotesCommandBase<CommandSourceStack, Ctx, Component>(
    ComponentFactory,
    CommandAdapter,
) {
    override fun performReloadConfig(): Boolean = mod.reloadConfig()

    fun register(dispatcher: CommandDispatcher<CommandSourceStack>) {
        val reload = Commands.literal("reload")
            .requires(
                mod.permissionsAdapter.permissionPredicate(
                    RELOAD_COMMAND_PERMISSION
                )
            )
            .executes(::tryReloadConfig)
        dispatcher.register(
            Commands.literal(NAMESPACE)
                .requires(
                    mod.permissionsAdapter.permissionPredicate(
                        COMMAND_PERMISSION
                    )
                )
                .then(reload)
        )
    }
}